### DISCUSSION TITLE
Getting Error while build source code in ubuntu20.04 :ignoring return value of 'fread'

### DISCUSSION THREAD

**Discussion by bokket** [pgmoneta/pgmoneta#110] [General]
Ref to [this iusse](https://github.com/gbdev/rgbds/issues/199).Different versions of libc and gcc complain in different ways. This need to add some code to check the returned values to make compiler happy.
This is the source problem:[#L286](https://github.com/pgmoneta/pgmoneta/blob/main/src/libpgmoneta/lz4_compression.c#L286) and [#L288](https://github.com/pgmoneta/pgmoneta/blob/main/src/libpgmoneta/lz4_compression.c#L288).
I don't know why have to read 4 bytes out first and then read from the `from` file to the `buffOut` buffer.These two items may solve.
1. Add return value to determine.It depends on the meaning of the different return values of `fread`.
2. Add `add_definitions(-w)` in `CMake` and it make it in my platfrom.

Also, the `starting-up` and `tutorial` on ubuntu is incomplete.

**Comment by jesperpedersen** [comment]
Can you send a pull request for this ?

If there is something wrong with the tutorials that is a separate discussion, but we should fix that as well.

**Reply by bokket** [reply]
Sorry, I didn't mean there is something wrong with the tutorials.
Because of the different platforms,the process of stepping in the pit makes me confused.

**Reply by jesperpedersen** [reply]
Is there something we can make clear ? We can create distribution specific sections if needed

**Reply by bokket** [reply]
1. Like make sure that pgmoneta is installed.Not using `pgmoneta -?` instead of `pgmoneta -v`.
2. While I have installed pgmoneta,its configure file:`pgmoneta.conf
` not in `/etc/pgmoneta` instead of `/usr/local/share/doc/pgmoneta/etc`,but run `pgmoneta` command need file under `/etc/pgmoneta` path.
3. `pgmoneta_users.conf`can't find,what steps did I miss?My postgreSQL have repl role:

```shell
                                   List of roles
 Role name |                         Attributes                         | Member of 
-----------+------------------------------------------------------------+-----------
 postgres  | Superuser, Create role, Create DB, Replication, Bypass RLS | {}
 repl      | Replication                                                | {}

```

**Reply by jesperpedersen** [reply]
1) Both commands will show if pgmoneta is installed correctly
2) We can't install in `/etc/pgmoneta/` plus we can't create a valid configuration for a user environment automatically - https://github.com/pgmoneta/pgmoneta/blob/main/doc/tutorial/01_install.md explains the steps
3) The first tutorial has this information - https://github.com/pgmoneta/pgmoneta/blob/main/doc/tutorial/01_install.md

**Reply by bokket** [reply]
Thanks for your help and patience.It seems to make it in my system:)
![深度截图_选择区域_20230315144655](https://user-images.githubusercontent.com/57066972/225230801-6f900c9c-9aa9-44fc-80d9-15b24dadf610.png)
I will add some information to prevent future ubuntu users from being plagued by path problems like me.

**Reply by jesperpedersen** [reply]
Check your pgmoneta log - it should have a backup if there is correct permissions for it

**Reply by bokket** [reply]
`cat > pgmoneta.conf`
```shell
...
log_type = file
log_level = info
log_path = /tmp/pgmoneta.log
...
```
I set it in the root directory. What's wrong with that?

**Reply by jesperpedersen** [reply]
And, you have the `[pgmoneta]` identifier ?

Did you work through the tutorials; especially in regards to the permissions on the PostgreSQL side ? Try and run `pg_basebackup` manually.

**Reply by bokket** [reply]
Hi.I understand what you mean.I did follow the tutorial and reply role has the permission on the PostgreSQL side.
```shell
                                  List of roles
 Role name |                         Attributes                         | Member of 
-----------+------------------------------------------------------------+-----------
 myuser    |                                                            | {}
 postgres  | Superuser, Create role, Create DB, Replication, Bypass RLS | {}
 repl      | Replication                                                | {}
```
The reason it didn't work before may be that the password for the primary user role`(pgmoneta side)` in the `pgmoneta.conf` must be the same as the repl role`(postgresql side)`.The pgmoneta.log is set in `/tmp/` by me.I found this iusse through the logs.
```shell
2023-03-17 10:37:13 INFO  main.c:621 pgmoneta: started on *
2023-03-17 10:37:13 WARN  security.c:1333 Wrong password for user: repl
2023-03-17 10:37:13 ERROR server.c:117 Server primary need wal_level at replica or logical
2023-03-17 10:37:13 ERROR wal.c:134 WAL: Server primary is not in a valid configuration
2023-03-17 10:37:35 ERROR backup.c:74 Backup: Server primary is not in a valid configuration
```
Maybe we need to mention this in the tutorial?Just use `ALTER ROLE repl WITH PASSWORD 'password';` to be consistent with `pgmoneta's master-key`(And `master-key` must enter more than 8 characters to enter successfully).

Then I retry to backup.It did make it this time.
```shell
Used space       : 16 MB
Free space       : 12.3 GB
Total space      : 100.0 GB
Link             : Yes
Number of servers: 1
Server           : primary
  Retention      : 7 days
  Backups        : 0
  Space          : 16 MB
```
The `WAL` file in `home/pgmoneta/backup` Try twice:
```shell
Used space       : 25.2 MB
Free space       : 12.3 GB
Total space      : 100.0 GB
Link             : Yes
Number of servers: 1
Server           : primary
  Retention      : 7 days
  Backups        : 1
                   20230317105508 (Backup: 7.9 MB Restore: 47.3 MB WAL: 32 MB Delta: 0 Retain: No Valid: Yes)
  Space          : 25.2 MB

```
I have come to understand the workflow of postgresql and will write two interesting proposals in the next week.
Any suggestions and discussions are welcomed!Thanks in advance!

**Comment by jesperpedersen** [comment]
> Maybe we need to mention this in the tutorial?Just use ALTER ROLE repl WITH PASSWORD 'password'; to be consistent with pgmoneta's master-key(And master-key must enter more than 8 characters to enter successfully).

The master key is shared for all users defined, and the user defined at the server level must have the password used to login into PostgreSQL. They should not be the same.

**Reply by bokket** [reply]
emmmm.... But the log warn:
```shell
2023-03-17 10:37:13 WARN  security.c:1333 Wrong password for user: repl
```
The first time I ran it, the password was not the same, and then I changed it to the same one.It puts 
```shell
2023-03-17 10:46:00 INFO  wal.c:120 WAL: primary
```

**Reply by jesperpedersen** [reply]
Yes, the user password has to be the same -- not the master key one

**Reply by bokket** [reply]
Yes.You are right!Here the tutorial is ambiguous, I believe the first time people will enter the same password as master-key as I did(pgmoneta side).This is a little naive.
As a side note:
1. pgmoneta does not support selective deletion of `configuration files and backup directory`.
2. If in the confige file `base_dir` is not exist,it should create.



I have a lot of trouble experimenting every time.
